src/binder/clrprivbindercoreclr.cpp

```diff
@@ -57,48 +58,58 @@ HRESULT CLRPrivBinderCoreCLR::BindAssemblyByName(IAssemblyName     *pIAssemblyNa
-                                                                   NULL, &pCoreCLRFoundAssembly);
-                if (SUCCEEDED(hr))
+                // If we are here, one of the following is possible:
+                //
+                // 1) The assembly has not been found in the current binder's application context (i.e. it has not already been loaded), OR
+                // 2) An assembly with the same simple name was already loaded in the context of the current binder but we ran into a Ref/Def
+                //    mismatch (either due to version difference or strong-name difference).
+                //
+                // Thus, if default binder has been overridden, then invoke it in an attempt to perform the binding for it make the call
+                // of what to do next. The host-overridden binder can either fail the bind or return reference to an existing assembly
+                // that has been loaded.
+
+                // Attempt to resolve the assembly via managed TPA ALC instance if one exists
+                INT_PTR pManagedAssemblyLoadContext = GetManagedAssemblyLoadContext();
+                if (pManagedAssemblyLoadContext != NULL)
                 {
-                    // We maybe returned an assembly that was bound to a different AssemblyLoadContext instance.
-                    // In such a case, we will not overwrite the binding context (which would be wrong since it would not
-                    // be present in the cache of the current binding context).
-                    if (pCoreCLRFoundAssembly->GetBinder() == NULL)
+                    tracer.Trace(BinderTracing::ResolutionAttemptedOperation::Stage::ALCLoad,
+                                 pManagedAssemblyLoadContext);
+
+                    hr = AssemblyBinder::BindUsingHostAssemblyResolver(pManagedAssemblyLoadContext, pAssemblyName, pIAssemblyName,
+                                                                       NULL, &pCoreCLRFoundAssembly);
+                    if (SUCCEEDED(hr))
                     {
-                        pCoreCLRFoundAssembly->SetBinder(this);
+                        // We maybe returned an assembly that was bound to a different AssemblyLoadContext instance.
+                        // In such a case, we will not overwrite the binding context (which would be wrong since it would not
+                        // be present in the cache of the current binding context).
+                        if (pCoreCLRFoundAssembly->GetBinder() == NULL)
+                        {
+                            pCoreCLRFoundAssembly->SetBinder(this);
+                        }
                     }
                 }
             }
-        }
 #endif // !defined(DACCESS_COMPILE) && !defined(CROSSGEN_COMPILE)
 
-        IF_FAIL_GO(hr);
-
-        *ppAssembly = pCoreCLRFoundAssembly.Extract();
+            tracer.SetResult(hr, pCoreCLRFoundAssembly);
```

Comment:
The assembly gets used in `ResolutionAttemptedOperation` to compute the error message - will `pCoreCLRFoundAssembly` actually be non-null for name mismatch and incompatible version failures?